File: C++/minimum-swaps-to-arrange-a-binary-grid.cpp
// Time:  O(n^2)
// Space: O(1)

class Solution {
public:
    int minSwaps(vector<vector<int>>& grid) {
        int result = 0;
        for (int target = grid.size() - 1; target >= 1; --target) {
            int row_idx = grid.size() - 1 - target;
            for (; row_idx < grid.size(); ++row_idx) {
                const auto& row = grid[row_idx];
                if (!accumulate(crbegin(row), crbegin(row) + target, 0)) {
                    break;
                }
            }
            if (row_idx == grid.size()) {
                return -1;
            }
            for (; row_idx != grid.size() - 1 - target; --row_idx) {
                swap(grid[row_idx], grid[row_idx-1]);
                ++result;
            }
        }
        return result;
    }
};

File: Python/minimum-swaps-to-arrange-a-binary-grid.py
# Time:  O(n^2)
# Space: O(1)

import itertools


class Solution(object):
    def minSwaps(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        result = 0
        for target in reversed(xrange(1, len(grid))):
            row_idx = len(grid)-1-target
            while row_idx < len(grid):
                row = grid[row_idx]
                if not sum(itertools.islice(row, len(row)-target, len(row))):
                    break
                row_idx += 1
            else:
                return -1
            while row_idx != len(grid)-1-target:
                grid[row_idx], grid[row_idx-1] = grid[row_idx-1], grid[row_idx]
                result += 1
                row_idx -= 1
        return result
